content/topics/sql-schema-design/index.md

```diff
@@ -0,0 +1,93 @@
+---
+title: SQL
+ready: True
+---
+
+# Database schema design
+
+
+In this topic, we will discuss the schema design. Here is what Wikipedia says about "Database Schema":
+
+> The database schema of a database is its structure described in a formal language supported by the database management system. The term "schema" refers to the organization of data as a blueprint of how the database is constructed. [Wikipedia](https://en.wikipedia.org/wiki/Database_schema)
+
+To help us understand the design process we will design an Umuzi.org physical database as an example. We will use Entity Relationship Diagram  (ERD) as a method to design our schema, I have selected [dbdiagram](https://dbdiagram.io) as a  tool to help us achieve this objective, I find the tool painless and straightforward, it provides fast feedback and I get to download a physical database right after, feel free to explore alternatives on the internet or experiment along.
+
+## The brief
+
+
+As a developer working on a product you have been given a ticket that has the following requirements:
+>Design a department database schema to handle data related to it.
+
+## The steps
+
+  Step 1:   idetify `entities` within a deparment
+  ---
+
+ Our "Department" schema to include the following entities:
+
+ - Departments
+ - Cohorts
+ - Recruits
+ - staffs
+![step 1 image](./img/step_1.png)
+
+
+  Step 2: Each attribute in a given entity should have a unique name
+  ---
+let establish all entities with their names and properties
+```
+ DEPARTMENT
+    -DEP_ID
+    -DEP_NAME
+
+COHORT
+    -COHORT_ID
+    -COHORT_NAME
+    -COHORT_START_DATE
+    -COHORT_END_DATE
+
+
+STAFF
+    -STAF_ID
+    -STAFF_TYPE
+    -STAF_FULL_NAME
+    -STAF_EMAIL
+    -STAF_DEPARTMENT
+
+RECRUIT
+    -RECRUIT_ID
+    -RECRUIT_FULL_NAME
+    -RECRUIT_DEPARTMENT
+    -RECRUIT-COHORT
+    -RECRUIT_PORTFOLIO_LINK
+```
+![step 2 image](./img/step_2.png)
+
+  Step 3: establish  relationships
+  ---
+
+  Each entity, attribute and relationship, should have appropriate names and that can be easily understood by the non-technical people as well. A relationship should not be connected directly to each other, a relationship should connect entities.
+
+ our entities within the "department schema" have the following relationships:
+
+- Each department has a one or more staff member, a staff member is part of only one department at a time
+- each cohort is linked to one or more department(s)
+- each cohort has one or more recruits
+
+We will consider only this scenario for now.
+
+![step 3 image](./img/step_3.png)
+
+Voila!
+
```

Comment:
Give them some examples of queries that will help them make sense of the structure.

Eg: 

If you know a recruit's full name you can get their cohort name like this: ...
And given a cohort you can get all the recruits in that cohort like this:...
This is a one to many relationship

And if you have a cohort's name you can get a list of all the department names associated with it like this ...  
And if you had a staff member then you can get a list of all the recruits within their department like this ...
These are called many-to-many relationships 

It would be good to demonstrate a one-to-one query as well

Reply:
Also it would be good to say that the data strucure is DRY.  Talk about what can go wrong if attributes are repeated. Eg:

Why dont we have a department_name in the recruits table? What would be the danget of doing this?